app/javascript/mastodon/components/emoji/context.tsx

```diff
@@ -0,0 +1,89 @@
+import type {
+  ComponentPropsWithoutRef,
+  ElementType,
+  PropsWithChildren,
+} from 'react';
+import {
+  createContext,
+  useCallback,
+  useContext,
+  useMemo,
+  useState,
+} from 'react';
+
+import classNames from 'classnames';
+
+import { cleanExtraEmojis } from '@/mastodon/features/emoji/normalize';
+import { autoPlayGif } from '@/mastodon/initial_state';
+import type {
+  CustomEmojiMapArg,
+  ExtraCustomEmojiMap,
+} from 'mastodon/features/emoji/types';
+
+// Animation context
+export const AnimateEmojiContext = createContext<boolean | null>(null);
+
+// Polymorphic provider component
+type AnimateEmojiProviderProps<Element extends ElementType = 'div'> =
+  ComponentPropsWithoutRef<Element> & {
+    as?: Element;
+    className?: string;
+  } & PropsWithChildren;
+
+export const AnimateEmojiProvider = ({
+  children,
+  as: Wrapper = 'div',
+  className,
+  ...props
+}: AnimateEmojiProviderProps<ElementType>) => {
+  const [animate, setAnimate] = useState(autoPlayGif ?? false);
+
+  const handleEnter = useCallback(() => {
+    if (!autoPlayGif) {
+      setAnimate(true);
+    }
+  }, []);
+  const handleLeave = useCallback(() => {
+    if (!autoPlayGif) {
+      setAnimate(false);
+    }
+  }, []);
```

Comment:
It'd probably be a good idea to forward the mouse event to `props.onMouseEnter/Leave` if they exist so they're not silently overridden. Or you just exclude those from the prop types :)